rsconnect/bundle.py

```diff
@@ -1245,6 +1439,90 @@ def write_notebook_manifest_json(
     return exists(join(directory, environment.filename))
 
 
+def create_voila_manifest(
+    path: str,
+    entrypoint: str,
+    environment: Environment,
+    app_mode: AppMode = AppModes.JUPYTER_VOILA,
+    extra_files: typing.List[str] = None,
+    excludes: typing.List[str] = None,
+    force_generate: bool = True,
+    image: str = None,
+    **kwargs
+) -> Manifest:
+    """
+    Creates and writes a manifest.json file for the given path.
+
+    :param path: the file, or the directory containing the files to deploy.
+    :param entry_point: the main entry point for the API.
+    :param environment: the Python environment to start with.  This should be what's
+    returned by the inspect_environment() function.
+    :param app_mode: the application mode to assume.  If this is None, the extension
+    portion of the entry point file name will be used to derive one. Previous default = None.
+    :param extra_files: any extra files that should be included in the manifest. Previous default = None.
+    :param excludes: a sequence of glob patterns that will exclude matched files.
+    :param force_generate: bool indicating whether to force generate manifest and related environment files.
+    :param image: the optional docker image to be specified for off-host execution. Default = None.
+    :return: the manifest data structure.
+    """
+    extra_files = list(extra_files) if extra_files else []
+    excludes = list(excludes) if excludes else []
+
+    entrypoint = entrypoint or infer_entrypoint(path=path, mimetype="text/ipynb")
+    deploy_dir = dirname(entrypoint)
+    manifest = Manifest(app_mode=AppModes.JUPYTER_VOILA, environment=environment, entrypoint=entrypoint, image=image)
+
+    if isfile(path):
+        validate_file_is_notebook(entrypoint)
+        manifest.entrypoint = entrypoint
+    else:
+        manifest.entrypoint = entrypoint or ""
+
+    # handle environment files
+    if not exists(join(deploy_dir, environment.filename)) or force_generate:
+        manifest.add_file(join(deploy_dir, environment.filename))
```

Comment:
We always need to add a `requirements.txt` file to the manifest file list, so this can't be conditional.

`environment.contents` contains the requirements file content, either from `requirements.txt` or `pip freeze`. You may need to create Manifest and Bundle methods to add a file buffer, as is done for the other content types, so that you can inject a requirements file into the manifest and bundle without writing it to the directory.

Reply:
fixed by [fcfbdd0](https://github.com/rstudio/rsconnect-python/pull/277/commits/fcfbdd0bbc418a7a881378981a797c579a8cd88a)

Reply:
now that I think about it, the original if condition is correct, because it prioritizes the user provided requirements.txt. Without it, we will always use the system generated one in environment.contents. I reverted the change in [cad1ae6](https://github.com/rstudio/rsconnect-python/pull/277/commits/cad1ae62a07b2b6326b79066bdf7d92044246c8a)

Reply:
We need to make sure these cases all work correctly:
* write-manifest with an existing requirements.txt
* write-manifest without requirements.txt
* deploy with an existing requirements.txt
* deploy without requirements.txt

I think that right now, the last one will not work because it won't include a requirements.txt in the bundle. You can unconditionally call
```
manifest.add_to_buffer(environment.filename, environment.contents)
```
because the environment inspection already prioritizes requirements.txt and will return its contents, if the file exists. It only invokes `pip freeze` if there is no requirements file.